src/plugins/intel_cpu/src/nodes/segment_max.cpp

```diff
@@ -0,0 +1,191 @@
+    try {
+        if (!ov::is_type<ov::op::v16::SegmentMax>(op)) {
+            errorMessage = "Only opset16 SegmentMax operation is supported";
+            return false;
+        }
+    } catch (...) {
+        return false;
+    }
+    return true;
+}
+
+void SegmentMax::getSupportedDescriptors() {
+    // Validation is already done in the ov::opset16::SegmentMax
+}
+
+void SegmentMax::initSupportedPrimitiveDescriptors() {
+    if (!supportedPrimitiveDescriptors.empty()) {
+        return;
+    }
+    ov::element::Type dataPrecision = getOriginalInputPrecisionAtPort(0);
+    ov::element::Type segmentIdsPrecision = getOriginalInputPrecisionAtPort(1);
+    if (getOriginalInputsNumber() == 2) {
+        addSupportedPrimDesc({{LayoutType::ncsp, dataPrecision}, {LayoutType::ncsp, segmentIdsPrecision}},
+                             {{LayoutType::ncsp, dataPrecision}},
+                             impl_desc_type::ref);
+    } else {
+        ov::element::Type numSegmentsPrecision = getOriginalInputPrecisionAtPort(2);
+        addSupportedPrimDesc({{LayoutType::ncsp, dataPrecision},
+                              {LayoutType::ncsp, segmentIdsPrecision},
+                              {LayoutType::ncsp, numSegmentsPrecision}},
+                             {{LayoutType::ncsp, dataPrecision}},
+                             impl_desc_type::ref);
+    }
+}
+
+bool SegmentMax::created() const {
+    return getType() == Type::SegmentMax;
+}
+
+bool SegmentMax::needPrepareParams() const {
+    return false;
+}
+
+void SegmentMax::executeDynamicImpl(const dnnl::stream& strm) {
+    execute(strm);
+
+    // Update lastSegmentIds
+    std::vector<int64_t> segmentIds(getSrcMemoryAtPort(1)->getSize());
+    auto segmentIdsPrecision = getOriginalInputPrecisionAtPort(1);
+    if (segmentIdsPrecision == ov::element::i32) {
+        const int32_t* srcSegmentIds = getSrcDataAtPortAs<const int32_t>(1);
+        std::copy(srcSegmentIds, srcSegmentIds + segmentIds.size(), segmentIds.begin());
+    } else if (segmentIdsPrecision == ov::element::i64) {
+        const int64_t* srcSegmentIds = getSrcDataAtPortAs<const int64_t>(1);
+        std::copy(srcSegmentIds, srcSegmentIds + segmentIds.size(), segmentIds.begin());
+    } else {
+        OPENVINO_THROW("Unsupported index type for segment_ids input");
+    }
+    lastSegmentIds.assign(segmentIds.begin(), segmentIds.end());
+
+    // Update lastNumSegments
+    if (getOriginalInputsNumber() == 3) {
+        const auto& numSegmentsPrecision = getOriginalInputPrecisionAtPort(2);
+        int64_t numSegmentsValue;
+        if (numSegmentsPrecision == ov::element::i64) {
+            numSegmentsValue = *getSrcDataAtPortAs<const int64_t>(2);
+        } else if (numSegmentsPrecision == ov::element::i32) {
+            numSegmentsValue = static_cast<int64_t>(*getSrcDataAtPortAs<const int32_t>(2));
+        } else {
```

Comment:
This part of the code has been added to fix incorrect ptr cast issue, but this change also bring back the question is it possible that i64 will reach this point?

Reply:
Asserted that it's always i32 and corrected.